Update overview, mission, scope, and roadmaps (#352)
neuroinformatics-unit/movement: docs/source/community/mission-scope.md

## Mission

`movement` aims to **facilitate the study of animal behaviour**
by providing a suite of **Python tools to analyse body movements**
across space and time.

## Scope

At its core, `movement` handles the positions of one or more individuals
tracked over time. An individual's position at a given time can be represented
in various ways: a single keypoint (usually the centroid), a set of keypoints
(also known as the pose), a bounding box, or a segmentation mask.
The spatial coordinates of these representations may be defined in 2D (x, y)
or 3D (x, y, z). The pose and mask representations also carry some information
about the individual's posture.

Animal tracking frameworks such as [DeepLabCut](dlc:) or [SLEAP](sleap:) can
generate these representations from video data by detecting body parts and

> tracking them across frames. In the context of `movement`, we refer to these trajectories as _tracks_: we use _pose tracks_ to refer to the trajectories of a set of keypoints, _bounding boxes' tracks_ to refer to the trajectories of bounding boxes' centroids, or _motion tracks_ in the more general case.

Reviewer: Note: add to #69 when merged

## Mission

`movement` aims to **facilitate the study of animal behaviour**
by providing a suite of **Python tools to analyse body movements**
across space and time.

## Scope

At its core, `movement` handles the positions of one or more individuals
tracked over time. An individual's position at a given time can be represented
in various ways: a single keypoint (usually the centroid), a set of keypoints

> (also known as the pose), a bounding box, or a segmentation mask.

Reviewer 2: > I added the "video data constraint", meaning that (at least for now) in movement we only consider data that is extracted from a video.

Do we need this constraint? Feasibly someone could track animals in a different way, and if they converted it to an appropriate file format, movement would work fine no?

> (also known as the pose), a bounding box, or a segmentation mask.

Author: I'm fine with your definition of "pose" here @sfmig, but I find it hard to write this section unambiguously.
The reason is that centroid tracking and (non-oriented) bboxes give no information about the animal's orientation, they just track its position, and `movement` can thus also represent position only, or position + some form of orientation.

The other confusing part for me is that if we use "pose" to mean "a set of keypoints on an individual" (which was my previous definition), that usually gives us many more things than just the "orientaton", e.g. we can often infer body length, shape, lengths and angle of particular body segments, etc. I guess this "narrower" definition of pose could be better described as capturing the "posture".

So if we are to be accurate about what `movement` represents, it would be something like: `position (required) + orientation (sometimes) + posture (sometimes)`.

> Do we need this constraint? Feasibly someone could track animals in a different way, and if they converted it to an appropriate file format, movement would work fine no?

That's right, we have discussed supporting trackball data, for example, and that doesn't come from a video. So though we are "primarily video-based" and "primarily pose estimation-focused", perhaps we shouldn't restrict ourselves here.

I think I have to think a bit more about this part, I'll come back with an alternative wording suggestion.

> (also known as the pose), a bounding box, or a segmentation mask.

Reviewer 3: > Do we need this constraint?

I'm fine doing away with the video constraint. I know we have discussed other options (we also talked about estimating pose from IMUs for [example](https://github.com/FIGLAB/IMUPoser)), but I thought we were keeping these out of scope for now. 

Maybe it is a good idea to widen the scope as we go? Meaning, if we have a use case in which we import, say translation and rotation per frame from IMU-derived data, then we expand the scope (as we are doing now). Just thinking out loud, I don't have strong views either way.

> (also known as the pose), a bounding box, or a segmentation mask.

Reviewer 3: > I find it hard to write this section unambiguously.

Yeah, I agree it is tricky. Initially I had written "position and/or orientation" instead of "pose" in the first sentence, but I felt it was cumbersome to read. I tried to cover the variability by expanding a bit on the examples ("could be _reduced_ to that of a single keypoint") but in truth I don't think a lot of people would say a centroid trajectory is a "pose". For me that level of small inconsistency in that paragraph is acceptable cause I find it is clarified by the examples.

> I guess this "narrower" definition of pose could be better described as capturing the "posture".

I would suggest not adding further new terms that are really not well-established concepts, such as "posture". I don't think many people think "posture" = "body length, shape, lengths and angle of particular body segments, etc". I think a better name for that would be "morphometrics". Without any other context, I would think "posture" = "pose" = "some representation of position and orientation."

> (also known as the pose), a bounding box, or a segmentation mask.

Author: > I would suggest not adding further new terms that are really not well-established concepts, such as "posture".
Agreed.

I've given this issues some more thought, and I personally think that defining "pose" in the more general, Computer Vision-y sense of "position an orientation" doesn't make sense right now, for the following reasons:
- Our API is already designed assuming that a "Pose" is a set of keypoints placed on an individual. That's why way make distinctions like `io.load_poses` vs `io.load_bboxes`, and also in the names of our datasets ("poses" vs "bboxes"). Usage of "poses" in that sense is peppered throughout our function names and docstrings. Therefore I don't think it makes sense to generalise that definition here in the scope, unless we commit to also redesigning the API accordingly.
- I had originally taken this meaning of "pose" and "pose tracking" from [this paper](https://www.nature.com/articles/s41593-020-00734-z) by Talmo et al., which is quite influential in neuro, and SLEAP follows this terminology. 

With that in mind, I've re-written the crucial section of scope in the following way (closely following @sfmig's nice structure but sticking with the "narrower" definition of "pose":

> At its core, `movement` handles the position and/or orientation of one or more individuals over time.
> 
> There are a few common ways of representing animal motion from video recordings: an animal's position could be reduced to that of a single keypoint tracked on its body (usually the centroid), or instead a set of keypoints (often referred to as the pose) to better capture its orientation as well as the postiions of limbs and appendages. The animal's position could be also tracked as a bounding box drawn around each individual, or as a segmentation mask that indicates the pixels belonging to each individual. Depending on the research question or the application, one or other format may be more convenient. The spatial coordinates of these representations may be defined in 2D (x, y) or 3D (x, y, z).
> 
> Animal tracking frameworks such as [DeepLabCut](dlc:) or [SLEAP](sleap:) can generate these representations from video data by detecting body parts and tracking them across frames. In the context of `movement`, we refer to these trajectories as _tracks_: we use _pose tracks_ to refer to the trajectories of a set of keypoints, _bounding boxes' tracks_ to refer to the trajectories of bounding boxes' centroids, or _motion tracks_ in the more general case.

Let me know what you think about this version. I'm still open to re-wording some sentences.

> (also known as the pose), a bounding box, or a segmentation mask.

Reviewer 3: I think this version looks pretty good. I would just make a tiny edit in the last paragraph from the quoted text:

> Animal tracking frameworks such as DeepLabCut or SLEAP can generate **keypoint** representations from video data by detecting body parts and tracking them across frames. In the context of movement, we refer to these trajectories as tracks: we use pose tracks to refer to the trajectories of a set of keypoints, bounding boxes' tracks to refer to the trajectories of bounding boxes' centroids, or motion tracks in the more general case.

I think specifying "keypoint" there makes it clear that these tools only refer to one of the representations mentioned above.

I agree with the other part about the definition of "pose". It is a bit unfortunate for newcomers or people less familiar with the problem that in the keypoint-based pose estimation world, the more general definition of pose has been equated to "a pose obtained with keypoints", and consequently to "keypoints". But maybe that is clear based on the context. We could make these clarifications about the use of "pose" in the glossary and link it here.

> unless we commit to also redesigning the API accordingly.

Maybe we end up here, if for example `load_poses` eventually also includes poses computed from ArUco markers. But maybe it is early to change the current `load_poses` to `load_keypoints` everywhere now, since at the moment we only support keypoint-based poses. We can revisit this when we support loading poses from other types of data.

The spatial coordinates of these representations may be defined in 2D (x, y)
or 3D (x, y, z). The pose and mask representations also carry some information
about the individual's posture.

Animal tracking frameworks such as [DeepLabCut](dlc:) or [SLEAP](sleap:) can

> generate these representations from video data by detecting positions and

Reviewer 2: ```suggestion
generate these representations from video data by detecting body parts and
```

> generate these representations from video data by detecting body parts and

Reviewer 3: ```suggestion
generate keypoint representations from video data by detecting body parts and
```

Reviewer 3: (but maybe I would consider removing this sentence entirely)

tracking them across frames. In the context of `movement`, we refer to the
resulting tracks according to their respective representations—for

> example, pose tracks, bounding boxes' tracks, or motion tracks in general.

Author: I think using "pose tracks" in this specific sense is quite common across papers etc., so I'm fine with it. I like how you've re-phrased this sentence in general.

As to which definition of "pose" we end up with, we are already having that discussion under the relevant comment.

Our vision is to present a **consistent interface for motion tracks** paired

> with **modular and accessible analysis tools**. We aim to accommodate data

Reviewer 3: ```suggestion
with **modular and accessible analysis tools**. We aim to support data
```

> Our vision is to present a **consistent interface for motion tracks** and to
> **analyze them using modular and accessible tools**. We aim to accommodate data

Reviewer 2: ```suggestion
Our vision is to present a **consistent interface for motion tracks** and
**modular and accessible analysis tools**. We aim to accommodate data
```

Author: How about this? ( I dislike multiple "and"s in a sentence).

> Our vision is to present a **consistent interface for motion tracks** paired with **modular and accessible analysis tools**.

Reviewer 2: Excellent

Reviewer 2: I mostly wanted to emphasise that we aren't doing the analysis, but providing a toolbox for others.

from a range of animal tracking frameworks, in **2D or 3D**, tracking
**single or multiple individuals**. As such, `movement` can be considered as

> downstream of tools like DeepLabCut and SLEAP. The focus is on providing

Reviewer 3: ```suggestion
operating downstream of tools like DeepLabCut and SLEAP. The focus is on providing
```

functionalities for data cleaning, visualization, and motion quantification
(see the [Roadmap](target-roadmaps) for details).

In the study of animal behavior, motion tracks are often used to extract and
label discrete actions, sometimes referred to as behavioral syllables or

> states. While `movement` is not designed for such tasks, it may generate

Reviewer 3: ```suggestion
states. While `movement` is not designed for such tasks, it can be used to generate
```

Author: "it can be used to generate features useful for" sounds awkard because of overusing "use" ;)

How about "While `movement` is not designed for such tasks, it may aid them by generating useful features and by inter-operating with existing action recognition tools".

Reviewer 3: Sure, I forgot to replace "useful".

The sentence "it may aid them" sounds a bit cumbersome to me, but maybe we can check with a native what sounds more natural. Some other options
- "it can be used to generate features that may be relevant for action recognition" 
- "it can be used to support the generation of features that may be informative for action recognition"

Reviewer: how about using the term "supplement"

Author: I like ""it can be used to generate features that may be relevant for action recognition".

Author: Thanks for the suggestions, both.

> features useful for action segmentation and recognition. We may develop

Reviewer 3: ```suggestion
features useful for action recognition. We may develop
```
to simplify, and avoid confusion with segmentation masks

> features useful for action segmentation and recognition. We may develop

Reviewer 3: ```suggestion
features useful for action segmentation and recognition. 
```
Should we remove this last sentence too? Feels a bit too broad.

Author: agreed. The last sentence doesn't belong here.
